lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,206 @@
+        to use Python 3 behaviour, but currently also works with Python 2. When
+        this is no longer required the TypeError exception and resulting action
+        can be removed.
+
+        Args:
+            attributes (dict):
+                A dictionary of attributes.
+            reference_attributes (dict):
+                A reference dictionary of attributes with which to make a
+                comparison.
+        Raises:
+            ValueError: Raised if the metadata extracted is not identical on
+                        all cubes.
+        """
+        try:
+            match = ((attributes.items() & reference_attributes.items()) ==
+                     reference_attributes.items())
+        except TypeError:
+            match = ((attributes.viewitems() &
+                      reference_attributes.viewitems()) ==
+                     reference_attributes.viewitems())
+        if match is not True:
+            raise ValueError('Cubes do not share the metadata identified '
+                             'by the grid_metadata_identifier ({})'.format(
+                                 self.grid_metadata_identifier))
+
+    def check_grid_match(self, cubes):
+        """
+        Uses the provided grid_metadata_identifier to extract and compare
+        attributes on the input cubes. The expectation is that all the metadata
+        identified should match for the cubes to be deemed compatible.
+
+        Args:
+            cubes (list of iris.cube.Cube items):
+                List of cubes for which the attributes should be tested.
+        """
+        # Allow user to bypass cube comparison by setting identifier to None.
+        if self.grid_metadata_identifier is None:
+            return
+
+        reference_attributes = self._get_attributes(cubes[0])
+        for cube in cubes:
+            attributes = self._get_attributes(cube)
+            self._compare_attributes(attributes, reference_attributes)
+
+    def process(self, spot_data_cube, neighbour_cube, gridded_lapse_rate_cube):
+        """
+        Extract lapse rates from the appropriate grid points and apply them to
+        the spot extracted temperatures.
+
+        The calculation is::
+
+         lapse_rate_adjusted_temperatures = temperatures + lapse_rate *
+         vertical_displacement
+
+        Args:
+            spot_data_cube (iris.cube.Cube):
+                The spot data cube of temperatures extracted from the gridded
+                fields. These temperatures will have been extracted using the
+                same neighbour_cube and neighbour_selection_method that are
+                being used here.
+            neighbour_cube (iris.cube.Cube):
+                The neighbour_cube that contains the grid coordinates at which
+                lapse rates should be extracted and the vertical displacement
+                between those grid points on the model orography and the spot
+                data sites actual altitudes.
```

Comment:
Is the neighbour cube a (generally) static ancillary (only changes when a new site is added)?  It would be good to state that somewhere.

Reply:
Added something to that effect.